rfcs/0046-versioning-changes-for-hard-fork.md

```diff
@@ -0,0 +1,152 @@
+## Summary
+[summary]: #summary
+
+Remove unneeded version tags from `Bin_prot`-serialized data produced
+by `ppx_version`. Generate version tags on an as-needed basis where
+they're useful. Detect versioning violations in the CI version linter
+in a more useful way.
+
+## Motivation
+[motivation]: #motivation
+
+Versioned data types are built from other versioned types and
+primitive OCaml types. When generating `Bin_prot`-serialized data from
+instances of those type, a version tag is generated not only for the
+containing type, but also for each constituent type. That strategy
+allows determining the version of any piece of the serialized data.
+
+The Jane Street `Async` versioned RPC mechanism used in
```

Comment:
How can our linting rules enforce that we are properly maintaining our async rpc version menu in accordance with our versioned types? Do those rules exist today?

Reply:
The RPC types have `deriving bin_io`; ppx_version enforces that the deriving list also has `version`; and the query and response types are not named `t`, so we have `version {rpc}` in the deriving list.

The version linter in CI is meant to catch changes to those query and response types. I created a dummy PR to see if that holds; the linter detected change to a response type, but not to a query type. I'll investigate.

Reply:
I worked out what's going wrong with the linter, will fix.

Reply:
Created #11121.